tests/test_crf.py

```diff
@@ -476,3 +476,186 @@ def assert_close(actual, expected):
+
+    def test_reduction_mean(self):
+        crf = make_crf()
+        crf_script = torch.jit.script(crf)
+
+        # shape: (seq_length, batch_size, num_tags)
+        emissions = make_emissions(crf)
+        # shape: (seq_length, batch_size)
+        tags = make_tags(crf)
+
+        llh = crf(emissions, tags, reduction='mean')
+        llh_script = crf_script(emissions, tags, reduction='mean')
+        assert_close(llh_script, llh)
+
+    def test_reduction_token_mean(self):
+        crf = make_crf()
+        crf_script = torch.jit.script(crf)
+
+        # shape: (seq_length, batch_size, num_tags)
+        emissions = make_emissions(crf)
+        # shape: (seq_length, batch_size)
+        tags = make_tags(crf)
+
+        mask = torch.tensor([[1, 1, 1], [1, 1, 0]], dtype=torch.uint8).transpose(0, 1)
+        llh = crf(emissions, tags, mask=mask, reduction='token_mean')
+        llh_script = crf_script(emissions, tags, mask=mask, reduction='token_mean')
+        assert_close(llh_script, llh)
+
+    def test_batch_first(self):
+        crf = make_crf()
+
+        # shape: (seq_length, batch_size, num_tags)
+        emissions = make_emissions(crf)
+        # shape: (seq_length, batch_size)
+        tags = make_tags(crf)
+
+        # Test scripted forward when running batch first mode
+        crf_bf = make_crf(batch_first=True)
+        # Copy parameter values from non-batch-first CRF; requires_grad must be False
+        # to avoid runtime error of in-place operation on a leaf variable
+        crf_bf.start_transitions.requires_grad_(False).copy_(crf.start_transitions)
+        crf_bf.end_transitions.requires_grad_(False).copy_(crf.end_transitions)
+        crf_bf.transitions.requires_grad_(False).copy_(crf.transitions)
+        crf_bf_script = torch.jit.script(crf_bf)
+        emissions = emissions.transpose(0, 1)
+        # shape: (batch_size, seq_length)
+        tags = tags.transpose(0, 1)
+        llh_bf = crf_bf(emissions, tags)
+        llh_bf_script = crf_bf_script(emissions, tags)
+        assert_close(llh_bf_script, llh_bf)
+
+
+@pytest.mark.skipif(
+    Version(torch.__version__) < Version("1.10.0"),
+    reason="torch version does not support torch script")
+class TestTorchScriptDecode:
+    def test_with_mask(self):
+        # Test decoding with a mask
+        crf = make_crf()
+        crf_script = torch.jit.script(crf)
+
+        seq_length, batch_size = 3, 2
+        # shape: (seq_length, batch_size, num_tags)
+        emissions = make_emissions(crf, seq_length, batch_size)
+        # mask should be (seq_length, batch_size)
+        mask = torch.tensor([[1, 1, 1], [1, 1, 0]], dtype=torch.uint8).transpose(0, 1)
+        best_tags = crf.decode(emissions, mask=mask)
+        best_tags_scripted = crf_script.decode(emissions, mask=mask)
+        assert best_tags == best_tags_scripted, f"scripted decode output {best_tags_scripted} " \
+                                                f"doesn't match non-scripted output {best_tags}"
```

Comment:
Why is `assert_close` not used here (and test functions below)?

Reply:
I thought that these are lists of integers that should not be compared approximately, but I forgot that I can control the tolerance. These are wrapped in a `assert_equal` function now to have the nice output of torch.testing.

Reply:
I just realised that they're integer lists! Using `assert` like this is preferable. Do you mind reverting the code back? But please remove the custom message as it might interfere with Pytest.

Reply:
The test were reverted to native comparison without the messages.